File: toy/fibonacci/FibonacciLastDigit.java
import java.util.*;

class FibonacciLastDigit
{
    static long fibonacci(int n, long n_1, long n_2)
    { 
      if (n == 0) return n_1;
      if (n == 1) return n_2;
      return fibonacci(n-1, n_2, Math.addExact(n_1, n_2));
    }
    
    static long fibonacciLastDigit(int n, long n_1, long n_2)
    { 
      if (n == 0) return n_1;
      if (n == 1) return n_2;
      return fibonacciLastDigit(n-1, n_2 % 10, (n_1 + n_2) % 10); // Math.addExact requires Java 8!
	}
     
    public static void main (String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        int n=0;
        System.out.print ("Enter a non-negative number: ");
        try {
			n = sc.nextInt();
			System.out.println("fib(" + n +") = " + fibonacci(n,0,1) );
		}
		catch (InputMismatchException e) {
			System.out.println ("Incorrect number.");
			System.exit(1);
		}
		catch (ArithmeticException e) {
			System.out.println ("The number n is too large");
		}
		
		try {
			System.out.println ("The last digit of nth fibonacci number is " + fibonacciLastDigit(n, 0, 1));
		}
		catch (StackOverflowError e) {
			System.out.println ("Too large a number to compute last digit tail recursively - Java has no optimization for tail calls.");
		}
		
    }
}

File: toy/fibonacci/fibonacci-lastdigit.rb
#!/usr/bin/env ruby

def fibonacci (n, n_1=0, n_2=1)
    begin
       if n==0 then n_1
       elsif n==1 then n_2
       else fibonacci(n-1, n_2, n_1+n_2); end
    rescue SystemStackError => e
       puts "fib(#{n}) is too large to compute"
    end
end

def fibonacci_lastdigit (n, n_1=0, n_2=1)
    begin
       if n==0 then n_1
       elsif n==1 then n_2
       else fibonacci_lastdigit(n-1, n_2 % 10, (n_1+n_2) % 10); end
    rescue SystemStackError => e
       puts "last digit of fib(#{n}) is too large to compute recursively because Ruby doesn't support tail recursion"
    end

end

def fibonacci_loop (n, only_last_digit=false) # this will support even huge numbers
    if (n==0) then return 0
    elsif (n==1) then return 1
    end
    n_2 = 0
    n_1 = 1
    result = 0
    for i in 2..n do
        result = only_last_digit==true ? (n_2 + n_1) % 10 : n_2 + n_1
        n_2 = n_1
        n_1 = result
    end
    result
end

printf "Enter a non-negative number: "
n = gets.to_i 
puts "fib(#{n}) = #{fibonacci(n, 0, 1)} (recursive-optimized) / #{fibonacci_loop(n)} (loop)"
puts "last digit of fib(#{n}) = #{fibonacci_loop(n, true)}"
